lib/dry/system/identifier.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+
+require "dry-equalizer"
+require_relative "constants"
+
+module Dry
+  module System
+    # An identifier representing a component to be registered.
+    #
+    # Components are eventually registered in the container using plain string
+    # identifiers, available as the `identifier` or `key` attribute here. Additional
+    # methods are provided to make it easier to evaluate or manipulate these identifiers.
+    #
+    # @api public
+    class Identifier
+      include Dry::Equalizer(:identifier, :namespace, :separator)
+
+      # @!attribute [r] identifier
+      #   @return [String] the identifier string
+      #   @api public
+      attr_reader :identifier
+
+      # @!attribute [r] namespace
+      #   @return [String, nil] the namespace for the component
+      #   @api public
+      attr_reader :namespace
+
+      # @!attribute [r] separator
+      #   @return [String] the configured namespace separator
+      #   @api public
+      attr_reader :separator
+
+      # @api private
+      def initialize(identifier, namespace: nil, separator: DEFAULT_SEPARATOR)
+        @identifier = identifier
+        @namespace = namespace
+        @separator = separator
+      end
+
+      # @!method key
+      #   Returns the identifier string
+      #
+      #   @return [String]
+      #   @see #identifier
+      #   @api public
+      alias_method :key, :identifier
+
+      # @!method to_s
+      #   Returns the identifier string
+      #
+      #   @return [String]
+      #   @see #identifier
+      #   @api public
+      alias_method :to_s, :identifier
+
+      # @!method to_str
+      #   Returns the identifier string
+      #
+      #   @return [String]
+      #   @see #identifier
+      #   @api public
+      alias_method :to_str, :to_s
```

Comment:
Why is this needed?

Reply:
It felt like the right thing to do for this class. Identifiers are strings at their core, so making them implicitly convertible to a string felt like the right behaviour. Means people working with an identifier could do things like `"foo " + component.identifier` and it'd just work.

You're right that it's not _strictly_ necessary, however. I'd be happy to remove it if you think it's a step too far. What do you think?

Reply:
@timriley if we want this to act as a string, we should have it as a decorator. IIRC `to_str` should be used when something quacks like a string and it can be coerced into a string when needed. Otherwise it'll be potentially surprising to the user that `start_with?` works but some other String methods do not.

Reply:
Fair point, I can remove it :)